_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,254 @@
+2) Generate TypeScript `.ts` output for working with Protobufs within your code
+2) Serialize messages into Protobuf for transmitting
+3) De-serialise Protobufs back in to useable message objects.
+
+For a next step, I would recommend spending spending some time understanding the [proto3 data structure](https://developers.google.com/protocol-buffers/docs/proto3) so that you can more easily build you own message structures.
+
+
+# How do do this so it looks like most published tutorials
+Most tutorials you find out there use a message class object with function calls to assign data to fields.  I have included this section for completeness and because I did go through this method before settling on my preferred solution above.   Particularly for data interchange you are likely to have a JSON object that you need to get into a Protobuf message and working through every data item and calling a function to assign to a field is tedious and results in more lines of code than necessary. Also, my comment above _*When you read the docs and get presented with ["The API is not well-documented yet"](https://github.com/protocolbuffers/protobuf/tree/master/js#api) you can be forgiven for wanting to cry and giving up_* reflects some of the pain I experienced when working through this method.
+
+So, very quickly:
+
+1) You need to use a different `protoc` plugin, for example [`ts-protoc-gen`](https://github.com/improbable-eng/ts-protoc-gen)
+2) CLI command to compile (note the package name is reversed after install which is dumb)
+```
+protoc \
+	--plugin="protoc-gen-ts=./node_modules/.bin/protoc-gen-ts" \
+	--js_out="import_style=commonjs,binary:./generated" \
+	--ts_out="./generated" \
+	src/proto/addressBook.proto
+```
+3) To create a new message
+    - Create a new instance of AddressBook:
+      - `const myAddressBook = new AddressBook()`
+    - You now have an empty address book, so add you first contact person:
+      - `const newPerson1 = myAddressBook.addPeople()`
+    - You now have one blank person in your address book, so add some contact details:
+      - `person1.setName('Joe Blogs')`
+      - `const person1Phone1 = person1.addPhones()`
+      - `person1Phone1.setNumber('0123456789')`
+      - `person1Phone1.setType(0)`
+        - Note that you have to use the index value of the enum for the `PhoneType`. 
+    - Now you get this idea, you can add the second person in the same way.
+4) Serialize to Protobuf:
+    - `const serializedMessage = myAddressBook.serializeBinary()`
+5) De-serialize back to message:
+    - `const deSerializedMessage = AddressBook.deserializeBinary(serializedMessage)`
+6) Get non-Typed object:
+    - `deSerializedMessage.toObject()`
+      - This one also caused me some trouble because without the Types and not knowing there was a `.toObject()` call, the `deSerializedMessage` object can be logged and is a mess of wrappers, nested arrays and no keys.
+
+Make you own choice which solution to use, there are advantages in each.
```

Comment:
we usually sign off with a way for the user to get in touch with us if they have any questions, or want to pay us loads of $$